std/experimental/allocator/mallocator.d

```diff
@@ -0,0 +1,226 @@
+module std.experimental.allocator.mallocator;
+import std.experimental.allocator.common;
+
+/**
+   The C heap allocator.
+ */
+struct Mallocator
+{
+    unittest { testAllocator!(() => Mallocator.instance); }
+
+    /**
+    The alignment is a static constant equal to $(D platformAlignment), which
+    ensures proper alignment for any D data type.
+    */
+    enum uint alignment = platformAlignment;
+
+    /**
+    Standard allocator methods per the semantics defined above. The
+    $(D deallocate) and $(D reallocate) methods are $(D @system) because they
+    may move memory around, leaving dangling pointers in user code. Somewhat
+    paradoxically, $(D malloc) is $(D @safe) but that's only useful to safe
+    programs that can afford to leak memory allocated.
+    */
+    @trusted void[] allocate(size_t bytes) shared
+    {
+        import core.stdc.stdlib : malloc;
+        if (!bytes) return null;
+        auto p = malloc(bytes);
+        return p ? p[0 .. bytes] : null;
+    }
+
+    /// Ditto
+    @system bool deallocate(void[] b) shared
+    {
+        import core.stdc.stdlib : free;
+        free(b.ptr);
+        return true;
+    }
+
+    /// Ditto
+    @system bool reallocate(ref void[] b, size_t s) shared
+    {
+        import core.stdc.stdlib : realloc;
+        if (!s)
+        {
+            // fuzzy area in the C standard, see http://goo.gl/ZpWeSE
+            // so just deallocate and nullify the pointer
+            deallocate(b);
+            b = null;
+            return true;
+        }
+        auto p = cast(ubyte*) realloc(b.ptr, s);
+        if (!p) return false;
+        b = p[0 .. s];
+        return true;
+    }
+
+    /**
+    Returns the global instance of this allocator type. The C heap allocator is
+    thread-safe, therefore all of its methods and `it` itself are
+    $(D shared).
+    */
+    static shared Mallocator instance;
+}
+
+///
+unittest
+{
+    auto buffer = Mallocator.instance.allocate(1024 * 1024 * 4);
+    scope(exit) Mallocator.instance.deallocate(buffer);
+    //...
+}
+
+unittest
+{
+    static void test(A)()
+    {
+        int* p = null;
+        p = new int;
+        *p = 42;
+        assert(*p == 42);
+    }
+
+    test!Mallocator();
+}
+
+unittest
+{
+    static void test(A)()
+    {
+        Object p = null;
+        p = new Object;
+        assert(p !is null);
+    }
+
+    test!Mallocator();
+}
+
+version (Posix) private extern(C) int posix_memalign(void**, size_t, size_t);
+version (Windows)
+{
+    private extern(C) void* _aligned_malloc(size_t, size_t);
```

Comment:
these functions (_aligned_malloc and its counterpart) don't exist in the DigitalMars standard C library (snn.lib). So currently DMD 32bit & OptLink are totally unable to link any code that would use these functions. and will produce

```
 Error 42: Symbol Undefined __aligned_free
 Error 42: Symbol Undefined __aligned_malloc
 Error 42: Symbol Undefined __aligned_realloc
```

the problem has already hit a few users of DCD when they tried to compile the sofware with -debug -gc.
Actually there is an underlying bug in DMD or optlink with this because when the external symbols are not used they are eliminated by the compiler (unless the switch -debug -gc are toogled on).

- https://github.com/Hackerpilot/DCD/issues/125
- http://forum.dlang.org/post/ueaabkpqlrpzdyemykkz@forum.dlang.org
- http://forum.dlang.org/post/fwnpahollypetjhqkjdw@forum.dlang.org

Maybe the error is that the author thought that either windows.lib or snn.lib contain these functions but **they don't**.
So under windows
- `version( DigitalMars )`: 
  - solution 1: either the `struct AlignedMallocator` has to be disabled.
  - solution 2: the functions have to be manually implemented, based on `malloc` (MS [states that it's based on malloc](https://msdn.microsoft.com/en-us/library/8z34s9c6.aspx), see remark section, so it should be do-able with snn malloc).
  - solution 3: @WalterBright  has to implement the functions in snn.
- `version( CRuntime_Microsoft )`: should be ok (DMD 64 bit).